### PR TITLE
script.xbmcbackup 1.6.1

### PR DESCRIPTION
**Pull request by robweber**
### Description
Update to the Backup addon compatible with Python 3. This has a couple of major changes vs the one in the repo currently (Krypton branch)

* Backups/Restores now use the concept of a Set to define groups of related files. Restores can restore one set or all sets within a backup archive (no more all or nothing restores)
* Added a new Advanced Editor script for more dynamic included/excluded directories based on a JSON formatted file
* Fixed guisettings restores
* Removed GoogleDrive support - Python 3 compatibility was an issue
* other fixes and formatting issues

__Please Note:__ There is a version of the Dropbox library bundled with this addon for Python 3 compatibility. Once the script.module.dropbox addon is updated this can and will be removed. I didn't want to delay a Python 3 version of this addon for those dependant on it. There is a PR submitted for the script.module.dropbox update to a Python 3 compatible version [here](https://github.com/micahg/script.module.dropbox/pull/14). 

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-scripts/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [script.foo.bar] 1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practise but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-scripts/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.

### CODE REVIEW

File: script.xbmcbackup/addon.xml [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    <description lang="sk_SK">Už ste niekedy poškodili konfiguráciu Kodi a priali si mať zálohu? Teraz môžete - na jeden klik. Môžete exportovať Vašu databázu, playlist, náhľady, doplnky a konfigurácie na ktorýkoľvek zdroj zapisovateľný Kodi. Zálohy môžu byť púšťané na požiadanie alebo plánovačom. </description>
+    <description lang="sv_SE">Har du någonsin tappat bort din Kodi konfiguration och önskat att du hade en backup? Nu kan du enkelt med ett klick. Du kan exportera din databas, spellista, minityrer, tillägg och andra konfigurationsdetaljer till valfri källa som är skrivbar för Kodi. Backupper kan köras på begäran eller via scheman.</description>
+    <description lang="zh_CN">你是否经常折腾你的 Kodi，因而希望能够有个备份？你可以把资料库、播放列表、缩略图、插件和其他配置细节导出到 Kodi 可以写入的任意位置。备份可以按需运行或通过计划任务执行。</description>
+    <language></language>
+    <platform>all</platform>
+    <license>The MIT License</license>
```
Comment (enen92) [inline script.xbmcbackup/addon.xml]: Can you use just `MIT`? (spdx identifier)

File: script.xbmcbackup/addon.xml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    <description lang="sk_SK">Už ste niekedy poškodili konfiguráciu Kodi a priali si mať zálohu? Teraz môžete - na jeden klik. Môžete exportovať Vašu databázu, playlist, náhľady, doplnky a konfigurácie na ktorýkoľvek zdroj zapisovateľný Kodi. Zálohy môžu byť púšťané na požiadanie alebo plánovačom. </description>
+    <description lang="sv_SE">Har du någonsin tappat bort din Kodi konfiguration och önskat att du hade en backup? Nu kan du enkelt med ett klick. Du kan exportera din databas, spellista, minityrer, tillägg och andra konfigurationsdetaljer till valfri källa som är skrivbar för Kodi. Backupper kan köras på begäran eller via scheman.</description>
+    <description lang="zh_CN">你是否经常折腾你的 Kodi，因而希望能够有个备份？你可以把资料库、播放列表、缩略图、插件和其他配置细节导出到 Kodi 可以写入的任意位置。备份可以按需运行或通过计划任务执行。</description>
+    <language></language>
```
Comment (enen92) [inline script.xbmcbackup/addon.xml]: please drop if empty

File: script.xbmcbackup/addon.xml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    <platform>all</platform>
+    <license>The MIT License</license>
+    <forum>https://forum.kodi.tv/showthread.php?tid=129499</forum>
+    <source>https://github.com/robweber/xbmcbackup</source>
+    <email></email>
```
Comment (enen92) [inline script.xbmcbackup/addon.xml]: plase drop if empty

File: script.xbmcbackup/default.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
+import sys
```
Comment (enen92) [inline script.xbmcbackup/default.py]: usually a good practice to keep this file to a minimum

File: script.xbmcbackup/resources/.gitignore [hunk 5]
```diff
@@ -0,0 +1,2 @@
+
```
Comment (enen92) [inline script.xbmcbackup/resources/.gitignore]: please drop

File: script.xbmcbackup/resources/data/default_files.json [hunk 6]
```diff
@@ -0,0 +1,105 @@
+{
```
Comment (enen92) [inline script.xbmcbackup/resources/data/default_files.json]: `__init__py` files should not be needed with python

File: script.xbmcbackup/resources/lib/advanced_editor.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+        exitCondition = ""
+        customPaths = BackupSetManager()
+
+        # show this every time
+        self.dialog.ok(utils.getString(30036), utils.getString(30037))
+
+        while(exitCondition != -1):
+            # load the custom paths
+            listItem = xbmcgui.ListItem(utils.getString(30126), '')
+            listItem.setArt({'icon': utils.addon_dir() + 'resources/images/plus-icon.png'})
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/advanced_editor.py]: use `os.path.join(utils.addon_dir() + 'resources','images','plus-icon.png')`

File: script.xbmcbackup/resources/lib/advanced_editor.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+            options = [listItem]
+
+            for index in range(0, len(customPaths.getSets())):
+                aSet = customPaths.getSet(index)
+
+                listItem = xbmcgui.ListItem(aSet['name'], utils.getString(30121) + ': ' + aSet['set']['root'])
+                listItem.setArt({'icon': utils.addon_dir() + 'resources/images/folder-icon.png'})
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/advanced_editor.py]: same

File: script.xbmcbackup/resources/lib/advanced_editor.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+                        if(self.dialog.yesno(heading=utils.getString(30127), line1=utils.getString(30128))):
+                            # delete this path - subtract one because of "add" item
+                            customPaths.deleteSet(exitCondition - 1)
+
+    def copySimpleConfig(self):
+        # disclaimer in case the user hit this on accident
+        shouldContinue = self.dialog.yesno(utils.getString(30139), utils.getString(30140), utils.getString(30141))
+
+        if(shouldContinue):
+            source = xbmc.translatePath(utils.addon_dir() + "/resources/data/default_files.json")
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/advanced_editor.py]: same

File: script.xbmcbackup/resources/lib/advanced_editor.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+            dest = xbmc.translatePath(utils.data_dir() + "/custom_paths.json")
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/advanced_editor.py]: same

File: script.xbmcbackup/resources/lib/backup.py [hunk 11]
```diff
@@ -0,0 +1,618 @@
...
+
+    def _createValidationFile(self, dirList):
+        valInfo = {"name": "XBMC Backup Validation File", "xbmc_version": xbmc.getInfoLabel('System.BuildVersion'), "type": 0}
+        valDirs = []
+
+        for aDir in dirList:
+            valDirs.append({"name": aDir['name'], "path": aDir['source']})
+        valInfo['directories'] = valDirs
+
+        vFile = xbmcvfs.File(xbmc.translatePath(utils.data_dir() + "xbmcbackup.val"), 'w')
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/backup.py]: xbmcvf now supports `contextlib`:

```python
with xbmcvfs.File(xbmc.translatePath(utils.data_dir() + "xbmcbackup.val"), 'w') as f:
     f.write()
```

File: script.xbmcbackup/resources/lib/backup.py [hunk 12]
```diff
@@ -0,0 +1,618 @@
...
+                    result = None
+
+        except ValueError:
+            # may fail on older archives
+            result = None
+
+        return result
+
+    def _createResumeBackupFile(self):
+        rFile = xbmcvfs.File(xbmc.translatePath(utils.data_dir() + "resume.txt"), 'w')
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/backup.py]: try to use contextlib if possible

File: script.xbmcbackup/resources/lib/backup.py [hunk 13]
```diff
@@ -0,0 +1,618 @@
...
+        rFile.write(self.restore_point)
+        rFile.close()
+
+    def _readBackupConfig(self, aFile):
+        jFile = xbmcvfs.File(xbmc.translatePath(aFile), 'r')
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/backup.py]: same

File: script.xbmcbackup/resources/lib/dropbox/auth.py [hunk 14]
```diff
@@ -0,0 +1,862 @@
+# -*- coding: utf-8 -*-
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/dropbox/auth.py]: maybe the dropbox section can be turned into a script.module?

File: script.xbmcbackup/resources/lib/tinyurl.py [hunk 15]
```diff
@@ -0,0 +1,12 @@
+# this is duplicated in snipppets of code from all over the web, credit to no one
+# in particular - to all those that have gone before me!
+from future.moves.urllib.request import urlopen
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/tinyurl.py]: Are you sure this works in python3?

File: script.xbmcbackup/resources/data/default_files.json (hunk 6 above)
Comment (robweber) [reply script.xbmcbackup/resources/data/default_files.json]: Didn't know these could be removed in Python 3 - thanks!

File: script.xbmcbackup/resources/lib/dropbox/auth.py (hunk 14 above)
Comment (robweber) [reply script.xbmcbackup/resources/lib/dropbox/auth.py]: Yes, in the description above I make mention of that fact that I do want to use the script.module.dropbox addon for this eventually. It is not currently at a version that supports Python3, I created a PR but still waiting. I can remove these in a future version once it's updated.

File: script.xbmcbackup/resources/lib/tinyurl.py (hunk 15 above)
Comment (robweber) [reply script.xbmcbackup/resources/lib/tinyurl.py]: Yes, I just tested it locally and it worked. Using a nightly build where the Python version is listed as 3.7.1.

File: script.xbmcbackup/resources/lib/dropbox/auth.py (hunk 14 above)
Comment (enen92) [reply script.xbmcbackup/resources/lib/dropbox/auth.py]: Where is the PR? You mean against our repo or against the upstream version?

**Comment by robweber** [comment]
I've updated the PR with the requested changes and responded to comments about others above. Thanks!

**Comment by enen92** [comment]
@robweber script.modules (specially those modules that come from pypi or similar repositories) that are shared without any patches are not tied to a particular author. Provider name is kept to the original author of the module, considering there aren't any changes in the code. Thus, I'd prefer if you could publish an update to repo-scripts instead of bundling the module here.

**Comment by robweber** [comment]
@enen92  - Understood, I recall having a similar discussion on the Kodi forum at one point as well. 

I'm attempting in this case to allow the author who usually maintains this module the opportunity to update (@micahg). My only reason for doing this is that the Dropbox module contains some non-Dropbox code, specifically these files: 

* [authenticate.py](https://github.com/micahg/script.module.dropbox/blob/master/authenticate.py)
* the [resources/ folder](https://github.com/micahg/script.module.dropbox/tree/master/resources)

I don't know the full story behind those additions or where else they might be used within other addons so I don't want to just pull them. If we need to sit on this until the Dropbox lib is merged in I can do that and update my PR accordingly. 

I do recall seeing some discussion that script.module addons should not contain additional code. If these additional files have no chance of being allowed in a future update I'm fine stripping them out and updating the lib myself.

**Comment by enen92** [comment]
@robweber if that's the case, the provider name for the script.module has to be changed from `Dropbox` to the respective author. I'll wait for the dropbox module PR.

**Comment by robweber** [comment]
I was referencing [this discussion](https://github.com/xbmc/repo-scripts/pull/1131#pullrequestreview-279324335) specifically from the last time the Dropbox module was updated. It was accepted with the current non-Dropbox code but sort of implied it wouldn't be again. Rechi's recommendation was to split them into a script.module for the Dropbox code and then another addon for the additions.

**Comment by micahg** [comment]
Hi @robweber and @enen92 -- I'm totally OK with you updating the addon. I'm a bit busy with other stuff at the moment if I'm being honest. Can you find the time?

**Comment by robweber** [comment]
I can do it, I guess I'd like clarification though on if it needs to be "pure" in order for it to be accepted. @enen92  can you comment on that? Do I need to remove the additional code pieces mentioned and then micahg can put them into their own addon later?

**Comment by enen92** [comment]
@robweber I'd appreciate that. Script modules, specially those from third parties should be as "pure" as possible and retain the author name. If that is not possible, the author name has to be changed so there is no confusion between both versions of the module (the one kodi is providing and the one in the third party site).
Be careful not to break any reverse dependencies because of the update.

**Comment by robweber** [comment]
Can do. I'm also waiting for the Dropbox module so I can pull that code from this request.

**Comment by enen92** [comment]
Dropbox module is merged, please adjust this one whenever possible

**Comment by robweber** [comment]
I updated the PR, this one has a different version number. 1.6.1 instead of 1.6.0. I did this for a few reasons:

1. I had already tagged the 1.6.0 version that included the Dropbox code so I wanted to differentiate between the version that had it included, and the one that used the module
2. Some bug fixes were added since the PR was made I wanted to also include in a new version
3. I modified how the file progress dialog works. Previously it used the file count as a progress indicator (filecount/total files). Now it uses the size of the data to be transferred (data transferred/total data size). This gives a more realistic idea of the time to do the backup as sometimes one file is huge compared to the others. 

I think I removed all the files you wanted gone (.gitignore, .travis.yml, etc). If I missed one let me know.

File: script.xbmcbackup/resources/.gitignore (hunk 5 above)
Comment (enen92) [inline script.xbmcbackup/resources/.gitignore]: still here

File: script.xbmcbackup/resources/lib/advanced_editor.py [hunk 16]
```diff
@@ -0,0 +1,233 @@
...
+            backupSet = {'name': name, 'root': rootFolder}
+
+        return backupSet
+
+    def editSet(self, name, backupSet):
+        optionSelected = ''
+        rootPath = backupSet['root']
+
+        while(optionSelected != -1):
+            options = [xbmcgui.ListItem(utils.getString(30120), "Exclude a specific folder from this backup set"), xbmcgui.ListItem(utils.getString(30135), "Include a specific folder to this backup set"), xbmcgui.ListItem(rootPath, utils.getString(30121))]
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/advanced_editor.py]: needs to be localized

File: script.xbmcbackup/resources/lib/advanced_editor.py [hunk 17]
```diff
@@ -0,0 +1,233 @@
...
+
+            for aDir in backupSet['dirs']:
+                if(aDir['type'] == 'exclude'):
+                    options.append(xbmcgui.ListItem(self._cleanPath(rootPath, aDir['path']), "%s: %s" % ("Type", utils.getString(30129))))
+                elif(aDir['type'] == 'include'):
+                    options.append(xbmcgui.ListItem(self._cleanPath(rootPath, aDir['path']), "%s: %s | %s: %s" % ("Type", utils.getString(30134), "Include Sub Folders", str(aDir['recurse']))))
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/advanced_editor.py]: needs to be localized

File: script.xbmcbackup/resources/lib/advanced_editor.py [hunk 18]
```diff
@@ -0,0 +1,233 @@
...
+                    self.dialog.ok(utils.getString(30117), utils.getString(30136), rootPath)
+            elif(optionSelected == 2):
+                self.dialog.ok(utils.getString(30121), utils.getString(30130), backupSet['root'])
+            elif(optionSelected > 2):
+
+                cOptions = ['Delete']
+                if(backupSet['dirs'][optionSelected - 3]['type'] == 'include'):
+                    cOptions.append('Toggle Sub Folders')
+
+                contextOption = self.dialog.contextmenu(cOptions)
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/advanced_editor.py]: needs to be localized

File: script.xbmcbackup/resources/lib/backup.py [hunk 19]
```diff
@@ -0,0 +1,637 @@
...
+    def _checkValidationFile(self, path):
+        result = None
+
+        # copy the file and open it
+        if(isinstance(self.remote_vfs, DropboxFileSystem)):
+            self.remote_vfs.get_file(path + "xbmcbackup.val", xbmc.translatePath(utils.data_dir() + "xbmcbackup_restore.val"))
+        else:
+            self.xbmc_vfs.put(path + "xbmcbackup.val", xbmc.translatePath(utils.data_dir() + "xbmcbackup_restore.val"))
+
+        vFile = xbmcvfs.File(xbmc.translatePath(utils.data_dir() + "xbmcbackup_restore.val"), 'r')
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/backup.py]: IIRC xbmcvfs in Matrix already supports contextlib:

```python
with xbmcvs.File(...) as vFile:
    do_something()
```

File: script.xbmcbackup/resources/lib/scheduler.py [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+
+        cron_ob = croniter(cron_exp, datetime.fromtimestamp(now))
+        new_run_time = cron_ob.get_next(float)
+
+        if(new_run_time != self.next_run):
+            self.next_run = new_run_time
+            utils.log("scheduler will run again on " + utils.getRegionalTimestamp(datetime.fromtimestamp(self.next_run), ['dateshort', 'time']))
+
+            # write the next time to a file
+            fh = xbmcvfs.File(self.next_run_path, 'w')
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/scheduler.py]: same

File: script.xbmcbackup/resources/lib/vfs.py [hunk 21]
```diff
@@ -0,0 +1,291 @@
...
+        return xbmcvfs.delete(aFile)
+
+    def rename(self, aFile, newName):
+        return xbmcvfs.rename(aFile, newName)
+
+    def exists(self, aFile):
+        return xbmcvfs.exists(aFile)
+
+    def fileSize(self, filename):
+        f = xbmcvfs.File(filename)
```
Comment (enen92) [inline script.xbmcbackup/resources/lib/vfs.py]: same

File: script.xbmcbackup/resources/.gitignore (hunk 5 above)
Comment (robweber) [reply script.xbmcbackup/resources/.gitignore]: :( I'm not sure why there is even an ignore file here, serves no purpose.

**Comment by enen92** [comment]
Happy new year :)

**Pull request merged**